File: 백준/Gold/10868. 최솟값/최솟값.py
import math
import sys

input = lambda: sys.stdin.readline().rstrip()


def build(node, start, end):
    if start == end:
        tree[node] = arr[start]
    else:
        mid = (start + end) // 2
        build(node * 2 + 1, start, mid)
        build(node * 2 + 2, mid + 1, end)
        tree[node] = min(tree[node * 2 + 1], tree[node * 2 + 2])


def query(node, start, end, left, right):
    # 범위 밖이라면
    if right < start or end < left:
        return math.inf

    # 전체가 범위 안이라면
    if left <= start and end <= right:
        return tree[node]

    mid = (start + end) // 2
    return min(query(node * 2 + 1, start, mid, left, right), query(node * 2 + 2, mid + 1, end, left, right))


if __name__ == '__main__':
    N, M = map(int, input().split())
    arr = [int(input()) for _ in range(N)]
    tree = [math.inf] * (4 * N)
    build(0, 0, N - 1)
    for _ in range(M):
        a, b = map(int, input().split())
        print(query(0, 0, N - 1, a - 1, b - 1))

File: 백준/Gold/10868. 최솟값/최솟값.cc
#include <bits/stdc++.h>
using namespace std;

int N, M;
vector<int> arr;
vector<int> tree;

void build(int node, int start, int end)
{
    if (start == end)
    {
        tree[node] = arr[start];
        return;
    }

    int mid = (start + end) / 2;

    build(node * 2, start, mid);
    build(node * 2 + 1, mid + 1, end);

    tree[node] = min(tree[node * 2], tree[node * 2 + 1]);
}

int query(int left, int right, int node, int start, int end)
{
    if (right < start || end < left)
    {
        return INT_MAX;
    }

    if (left <= start && end <= right)
    {
        return tree[node];
    }

    int mid = (start + end) / 2;
    return min(query(left, right, node * 2, start, mid), query(left, right, node * 2 + 1, mid + 1, end));
}

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);

    cin >> N >> M;
    arr.resize(N);
    tree.resize(4 * N);

    for (int i = 0; i < N; i++)
    {
        cin >> arr[i];
    }

    build(1, 0, N - 1);

    for (int i = 0; i < M; i++)
    {
        int a, b;
        cin >> a >> b;
        cout << query(a - 1, b - 1, 1, 0, N - 1) << "\n";
    }

    return 0;
}